rclcpp/topics/minimal_publisher/member_function_type_masquerading.cpp

```diff
@@ -0,0 +1,57 @@
+// Copyright 2016 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <functional>
+#include <memory>
+#include <string>
+
+#include "rclcpp/rclcpp.hpp"
+#include "std_msgs_conversions/std_string.hpp"
+
+using namespace std::chrono_literals;
+
+/* This example creates a subclass of Node and uses std::bind() to register a
+ * member function as a callback from the timer. */
+
+class MinimalPublisher : public rclcpp::Node
+{
+public:
+  MinimalPublisher()
+  : Node("minimal_publisher"), count_(0)
+  {
+    publisher_ = this->create_publisher<std::string>("topic", 10);
+    timer_ = this->create_wall_timer(
+      500ms, std::bind(&MinimalPublisher::timer_callback, this));
+  }
+
+private:
+  void timer_callback()
+  {
+    std::string message = "Hello, world! " + std::to_string(count_++);
+    RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.c_str());
+    publisher_->publish(message);
+  }
+  rclcpp::TimerBase::SharedPtr timer_;
+  rclcpp::Publisher<std::string>::SharedPtr publisher_;
```

Comment:
I know we discussed this in the past, but another option here is to be more explicilt and say what `std::string` is a stand in for.

That would have the benefit of being more explicit to read and also would potentially allow us to have `std::string` map to more than one ROS type, for example.

Ideas for how that might look:

```c++
rclcpp::Publisher<std::string, std_msgs::msg::String>::SharedPtr publisher_;
```

Or the reverse order might make sense:

```c++
rclcpp::Publisher<std_msgs::msg::String, std::string>::SharedPtr publisher_;
```

But that tells me this is actually kind of ambiguous, so maybe some machinery to make it clear (though less concise):

```c++
rclcpp::Publisher<Masquerade<std::string>::as<std_msgs::msg::String>>::SharedPtr publisher_;
```

Again the order of the type could be reversed and kind of make sense? Maybe this is clearer:

```c++
rclcpp::Publisher<Use<std::string>::InsteadOf<std_msgs::msg::String>>::SharedPtr publisher_;
```

I dunno, the naming can be tweaked, but in general having some syntax to make it clear what's happening might help. We could always support both the concise and verbose options too I guess.

Reply:
Talking with @audrow, we thought:

- Call it something else? Masquerade is a weird word, might be harder for non-English (non-French) speakers.
  - Facade (also French, but more common) makes more sense? It's shorter and the metaphor for Facade is better since it has roots in architecture.
- How should it be stated? `Facade<std::string>::instead_of<std_msgs::msg::String>` or `Facade<std_msgs::msg::String>::instead_use<std::string>`?
  - We think `Facade<std::string>::instead_of<std_msgs::msg::String>` because the "facade" is the `std::string`, and it reads better. Also the one we're "using" is farther to the left and so you see it first.
- Should we support more than one type mapping per user type?
  - It would be useful, but makes the implicit declaration (`Publisher<std::string>`) more difficult because it would only work if there's one ROS type that `std::string` could represent.

How might it work:

```c++
template<class From, class To>
struct TypeMapping
{
  // ...
};

template<class FacadeType>
struct TypeFacade {
  template<class InterfaceType>
  using instead_of = TypeMapping<FacadeType, InterfaceType>;
};

// you can be explicit like this:
using MsgT = rclcpp::TypeFacade<std::string>::instead_of<std_msgs::msg::String>;
rclcpp::Publisher<MsgT>::SharedPtr publisher_;
// ^ you would use this if you want it to be clear what is happening
//   (maybe examples or depending on the preferences of the developer), or
//   you might use it when `std::string` could map to more than one ROS type
//   and you need to select which one to use.

// We could have a short form of this that takes two arguments:
rclcpp::Publisher<TypeMapping<std::string, std_msgs::msg::String>>::SharedPtr publisher_;
// Or this, but it might be prohibitively complicated to implement
// since there are multiple template args already for Publisher/Subscription:
rclcpp::Publisher<std::string, std_msgs::msg::String>::SharedPtr publisher_;

// or you can be implicit:
rclcpp::Publisher<std::string>::SharedPtr publisher_;
// ^ If std::string has multiple types it could represent then this has to error.
```

Reply:
> Call it something else? Masquerade is a weird word, might be harder for non-English (non-French) speakers.

the technique is called *type masquerading*. It seems a bit artificial to try and find a different name for it in ROS 2, especially since there has been major effort in its design to use standard(ised) technologies and approaches in the overall design. Unless there is a really good reason not to use the established name for this, I would recommend against it.

Additionally, *facading* is not a verb, while masquerading is. Minor, but makes it easier to discuss it.

The code sample reads a bit "forced" as well: 

```c++
Facade<...>::instead_of<...>
```

is not nearly as easy to grok as:

```c++
rclcpp::Publisher<Masquerade<...>::as<...>>::SharedPtr publisher_;
```

which communicates the *use-this-instead-of-that* intent very nicely.

Reply:
Thanks, we'll take that into consideration, and we're planning on making a REP and listing this in the alternative section (which ever one we do not select for the name).

> the technique is called type masquerading. It seems a bit artificial to try and find a different name for it in ROS 2,

When @audrow mentioned to me that the name might be confusing and we should consider alternatives, I was open to it mostly because AFAIK I had just invented the term `type masquerading` years ago. It's possible I didn't and instead I heard it from some ROS 1 stuff, but recently when I was spinning back up on this topic and looking at ROS 1's documentation on this, they don't call it that at all:

http://wiki.ros.org/roscpp/Overview/MessagesSerializationAndAdaptingTypes#Adapting_C.2B-.2B-_Types

It's referred to as "adapting C++ types", at least in the documentation.

I remember taking inspiration from IP Masquerading, which is a somewhat related concept to do with NAT routing, but other than that I'm not aware of a broader CS technique about `type masquerading`.

After talking about it `facade` was interesting for the reasons I noted above.

That all being said I actually don't mind what's called, and I take the point that you think `Masquerade<...>::as<...>` is clearer.

Just out of curiosity which do you think is more correct @gavanderhoorn?

```c++
Masquerade<std_msgs::msg::String>::as<std::string>
// or
Masquerade<std::string>::as<std_msgs::msg::String>
```

> Additionally, facading is not a verb (it is a band apparently :) ), while masquerading is. Minor, but makes it easier to discuss it.

I suppose so, but I was thinking of "a Type Facade", and in the code `TypeFacade<X>` is more correct in an object oriented sense, as `Masquerade<X>` reads a both a noun and a verb and as a noun it doesn't make sense to me. Again, it doesn't really matter to me.

Reply:
>  > the technique is called type masquerading. It seems a bit artificial to try and find a different name for it in ROS 2,
> 
> When @audrow mentioned to me that the name might be confusing and we should consider alternatives, I was open to it mostly because AFAIK I had just invented the term type masquerading years ago.
> 
> I remember taking inspiration from IP Masquerading, which is a somewhat related concept to do with NAT routing, but other than that I'm not aware of a broader CS technique about `type masquerading`.

So I think you may be right actually. I've heard the term being used around me a lot, but perhaps that's in a different context. IP masquerading is indeed a thing, as is a masquerade attack (pretending to be an authorised user). There are some other contexts in which the term is used in a similar fashion, but I can't find an official(-ish) reference for you specifically dealing with *types* in the sense that we're discussing here.

So to me, it's completely clear what is meant by *type masquerading*. I'd also assume that for anyone familiar with facades, proxies, wrappers or similar constructs, it would not difficult to infer the meaning.

But if you don't know the connection between mask and masquerade, then it could be difficult to figure out.

That said though: I believe the term does work for the intended purpose.

> Just out of curiosity which do you think is more correct @gavanderhoorn?
> 
> ```
> Masquerade<std_msgs::msg::String>::as<std::string>
> // or
> Masquerade<std::string>::as<std_msgs::msg::String>
> ```

so if "to masquarade" is intended to mean: *pretend to be something/someone else*, then only the second option would seem to be correct, and would mean: take `std::string`, and anywhere you find me using a `std::string` (in specific contexts) pretend I wrote `std_msgs::msg::String` instead.

Which seems to be what you're after, right?

---

Thinking about it some more: `Pretend<..>::is_a<..>` could be an alternative.

Reply:
Again, we'll take that into consideration, but if we continue this discussion maybe on the new draft pr for the rep is a better place to discuss it:

https://github.com/ros-infrastructure/rep/pull/303

We have a section there where we are discussing naming and alternatives considered.